### ISSUE TITLE
Contending with multiple data sources

### ISSUE THREAD

**Issue by dlandtaa** [dariosalvi78/cordova-plugin-health#281]
We have an app that is reading various items from Apple Health and Google Fit. When we have other apps installed on those devices such as Garmin Connect that also write to Health and Fit we see that data show up in the queries results along with the data natively collected by the phone itself. A lot of this data is duplicated by both phone and Garmin. For example both the phone and the Garmin watch are recording sleep data and writing it to Apple Health and you end up with a lot of duplicate information:

<img src="https://user-images.githubusercontent.com/89596857/228999960-333a5dc2-6e0f-4204-81d7-a60a68425cfb.jpeg" width="400">

So just getting the sum of all the sleep data for a particular day is going to give you a number that's way too high.

So I think what I'd like to do is use only the data created by the phone, not by Garmin. I was looking at the `sourceBundleId` field for this. On my iPhone that value is `com.apple.health.1B4A8376-F409-4122-879C-66CB42F36579`. I don't know what the UUID like thing at the end is. Is that specific to my device? Is it valid to look for all entries where `sourceBundleId` starts with `"com.apple.health"`? Garmin entries appear to have a `sourceBundleId` of `"com.garmin.connect.mobile"`. I noticed that data entered directly into the Health app manually seem to use the bundle id `"com.apple.Health"` which I think must be the id of the Health app. 

It's a similar situation on Android but the `sourceBundleId` is something like `"raw:com.google.step_count.cumulative:Razer:Phone:50f6f5ad:Step Counter"` for steps recorded automatically by the phone. So is it valid to look for entries that start with `"raw:com.google.step_count"` or maybe just `"raw:com.google"`? For manual entries it seems to use `"com.google.android.apps.fitness"`.

Of course there are probably other apps that write to Health and Fit bedsides Garmin. We can't control what our users have installed on their phone so we need to be able to find just the data created by the device itself.

**Comment by dariosalvi78** [comment]
this is a good question and I have no answer for it!
It looks like the best thing to do is to look at sourceBundleIds and find commonalities, exactly as you are doing. There may be ways to add filters to the queries in native code, but so far I haven't seen anything really spot-on ([example for Fit](https://developers.google.com/android/reference/com/google/android/gms/fitness/request/DataReadRequest.Builder)).

I can only recommend that if you do your analysis, please tell us what your conclusions are, so that we can all benefit.

Thank you!

**Comment by dariosalvi78** [comment]
actually one way could be to aggregate by data source, see [this](https://developers.google.com/android/reference/com/google/android/gms/fitness/request/DataReadRequest.Builder#aggregate(com.google.android.gms.fitness.data.DataSource)). This would work for aggregated queries, and we would need to see how this plays with the added datatype aggregation. We would also need to see if what the equivalent for HealthKit would be.

It would definitely be an interesting addition to the plugin, but it's one that requires quite some work. If anybody is up for the job, you are very welcome!

**Comment by dlandtaa** [comment]
Okay, thank you for the info. I have not been able to successfully get the garmin app to write to google fit so I'm guessing it doesn't have that capability. Do you know off hand of any devices that can write data to google fit?

**Comment by dariosalvi78** [comment]
I think that the Xiaomi app can write to Google Fit. Xiaomi Bands are also very cheap, definitely worth trying.

**Comment by dlandtaa** [comment]
So I was l playing with the aggregate queries and noticed that when I use `queryAggregated()` for steps it actually gives me the "merged" total displayed in the Health app. I had data from my phone, my apple watch, and my garmin watch and it seemed to combine them all together using whatever algorithm apple uses. I didn't realize that was part of the aggregation process.

Unfortunately when I tried this with sleep it said that data type cannot be aggregated. Not sure why that is.

**Comment by dlandtaa** [comment]
Just wanted to update this in case it's helpful to anyone else.

The way we addressed this was to use aggregation for steps and to come up with our own algorithm for handling sleep data. As far as I know, Apple doesn't make the algorithm that Health uses to merge different data sources together, but I was able to kind of get the gist by experimenting a bit. It looks like any overlapping sleep intervals are merged together into one long interval. So if your Apple watch says you slept from 11:00 pm to 7:00 am and your Garmin watch says you slept from midnight to 8:00 am, the Health app will log that as having sleep from 11:00 pm to 8:00 am. Not perfect but good enough for us, so we basically wrote our own version of that.

I also noticed that it seems to favor sleep data with a value of "sleep" other "sleep.inBed" (which is what the iPhone logs it as) so we did the same. If there is any intervals of type "sleep" it will use those and ignore the others. If not it will use any "sleep.inBed" values that are available.

Google has a different approach. They seem to prioritize one data source over another and use just data from that one if it exists. If I let the Sleep Cycle app on my phone record some sleep data and then I manually enter some sleep data into the Fit app that overlaps the Sleep Cycle interval, it will totally ignore the Sleep Cycle data and use my manually entered data. That's a way to go, but we didn't do that.

For other metrics like HRV that are discrete values that aren't really aggregated like steps or sleep we just let the different sources add their data and we use it all.
